dbms/src/Functions/FunctionsIsIPAddr.h

```diff
@@ -0,0 +1,274 @@
+                        return 0;
+                    colonp = tp;
+                    continue;
+                }
+                if (tp + INT16SZ > endp)
+                    return 0;
+                *tp++ = static_cast<unsigned char>(val >> 8) & 0xff;
+                *tp++ = static_cast<unsigned char>(val) & 0xff;
+                saw_xdigit = 0;
+                val = 0;
+                continue;
+            }
+            if (ch == '.' && ((tp + INADDRSZ) <= endp) && CheckIsIPv4Impl::isMatch(curtok) > 0)
+            {
+                tp += INADDRSZ;
+                saw_xdigit = 0;
+                break; /* '\0' was seen by CheckIsIPv4Impl::isMatch(). */
+            }
+            return 0;
+        }
+        if (saw_xdigit)
+        {
+            if (tp + INT16SZ > endp)
+                return 0;
+            *tp++ = static_cast<unsigned char>(val >> 8) & 0xff;
+            *tp++ = static_cast<unsigned char>(val) & 0xff;
+        }
+        if (colonp != nullptr)
+        {
+            const size_t n = tp - colonp;
+            size_t i;
+
+            for (i = 1; i <= n; ++i)
+            {
+                endp[-i] = colonp[n - i];
+                colonp[n - i] = 0;
+            }
+            tp = endp;
+        }
+        if (tp != endp)
+            return 0;
+        return 1;
+    }
+};
+
+template <typename Impl>
+class FunctionIsIPv4OrIsIPv6 : public IFunction
+{
+public:
+    static constexpr auto name = Impl::name;
+    FunctionIsIPv4OrIsIPv6() = default;
+
+    static FunctionPtr create(const Context &) { return std::make_shared<FunctionIsIPv4OrIsIPv6>(); };
+
+    std::string getName() const override { return name; }
+    size_t getNumberOfArguments() const override { return 1; }
+    bool useDefaultImplementationForConstants() const override { return true; }
+    bool useDefaultImplementationForNulls() const override { return false; }
+
+    DataTypePtr getReturnTypeImpl(const DataTypes & arguments) const override
+    {
+        if (arguments.size() != 1)
+            throw Exception(
+                fmt::format("Number of arguments for function {} doesn't match: passed {}, should be 1.", getName(), arguments.size()),
+                ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH);
+        if (!arguments[0]->onlyNull())
+        {
+            DataTypePtr data_type = removeNullable(arguments[0]);
+            if (!data_type->isString())
+                throw Exception(
+                    fmt::format("Illegal argument type {} of function {}, should be integer", arguments[0]->getName(), getName()),
+                    ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+        }
+        return std::make_shared<DataTypeUInt8>();
+    }
+    void executeImpl(Block & block, const ColumnNumbers & arguments, size_t result) const override
+    {
+        auto [column, nullmap] = removeNullable(block.getByPosition(arguments[0]).column.get());
+        if (const auto * col_input = checkAndGetColumn<ColumnString>(column))
```

Comment:
Does this hold true for the case where the input is `only null`?
Is there an ut to guarantee this?

Reply:
thank u very much~

Before there was a test case where the input was all `NULL` for testing `ConstColumn`
https://github.com/AntiTopQuark/support_ipv4_and_ipv6_func/blob/09fd7fd29c83f298ee654c9b9a416601a718b854/dbms/src/Functions/tests/gtest_is_ip_addr.cpp#L80
```
    ASSERT_COLUMN_EQ(createConstColumn<UInt8>(4, 0), executeFunction("tiDBIsIPv6", {createConstColumn<Nullable<String>>(4, std::nullopt)}));
```
now I added another one
```
    ASSERT_COLUMN_EQ(createColumn<UInt8>({0, 0, 0, 0, 0}), executeFunction("tiDBIsIPv6", {createColumn<Nullable<String>>({std::nullopt, std::nullopt, std::nullopt, std::nullopt, std::nullopt})}));

```

Reply:
Can you test for `DataTypeNullable<DataTypeNothing>`?
You can use 
```
ColumnWithTypeAndName createOnlyNullColumnConst(size_t size, const String & name = "");
ColumnWithTypeAndName createOnlyNullColumn(size_t size, const String & name = "");
```
in `FunctionTestUtils`.

Reply:
You can refer to `gtest_strings_concat_ws.cpp`